Initial compatibility for unprivileged Google Play Services

kdrag0n opened the pull request:
This makes it possible to run Google Play Services (aka. Google Mobile Services / "GMS") and related apps (Google Services Framework, Google Play Store) as regular, unprivileged user apps.

Tested functionality:
  - Account login (including 2FA with NFC security key)
  - Play Store
  - Play Store app installation
  - Firebase Cloud Messaging (Signal, Discord, Slack, Gmail)
  - Firebase database API (Swift Backup)
  - Firebase app indexing (GMS debug settings)
  - Google Play Games
  - Account settings
  - Google My Account
  - Autofill
  - SMS verification receiver (Signal)
  - Play license verification (both in-app purchases and paid apps)
  - Play Store app purchases

Known issues:
  - Play feature delivery
  - Play Store app installation gets stuck on "Installing..." even when the installation succeeds
  - Chrome fails to install
  - Google Camera app isn't compatible according to Play Store
  - Doesn't work in secondary users

To open Play Services and account settings: `am start com.google.android.gms/.app.settings.GoogleSettingsIALink`

empratyush committed:
Bluetooth auto turn off

empratyush committed:
Wi-Fi auto turn off

kdrag0n committed:
gmscompat: Add helpers for Google Play Services compatibility

This adds a set of helper classes and methods for our upcoming Google
Play Services compatibility, which will make it possible to run Google
Play Services (aka. Google Mobile Services / "GMS") and related apps
(Google Services Framework, Google Play Store) as regular, unprivileged
user apps.

Implementation notes:
  - The core GmsCompat class is in android.app.gmscompat and exposes
    isEnabled() to system modules because WifiManager, which is part of
    the com.android.wifi APEX module, will need to use it.
  - GmsCompat should only be used for core functionality. API shim
    implementations belong in GmsHooks.
  - The flag has been implemented as a compatibility change,
    GMS_UNPRIVILEGED_COMPAT, in order to reuse the existing
    compatibility change infrastructure.

The actual API shims and compatibility will be added in following
commits.

Change-Id: Ibfabb6465c40280d4ce73c120ff1f18230fbe2d3

kdrag0n committed:
gmscompat: Keep GMS services alive by converting to foreground services

Unprivileged apps can't be exempt from the background execution
requirements, so keep GMS services alive by converting them to
foreground services.

Apps must post a foreground service notification for each service within
10 seconds of the service starting or the system will kill the entire
process, so generate notifications dynamically from fully-qualified
service class names and post them automatically for GMS apps.

Change-Id: I0c8a44bf86c364dbbd22ef18138a9a57c86a7bdc

kdrag0n committed:
gmscompat: Add baseline compatibility for unprivileged Play Services

Tested functionality:
  - Account login (including 2FA with NFC security key)
  - Play Store
  - Firebase Cloud Messaging (Signal, Discord, Slack, Gmail)
  - Firebase database API (Swift Backup)
  - Firebase app indexing (GMS debug settings)
  - Google Play Games
  - Account settings
  - Google My Account
  - Autofill
  - SMS verification receiver (Signal)
  - Play license verification (both in-app purchases and paid apps)
  - Play Store app purchases

Known issues:
  - Play feature delivery
  - SafetyNet doesn't pass

To open Play Services and account settings:
      am start com.google.android.gms/.app.settings.GoogleSettingsIALink

Change-Id: I80c702ac47590d28c86cd1a63a7765670ec9e045

kdrag0n committed:
gmscompat: Enable compatibility change for GMS apps

This overrides the default disabled state for the compatibility change
and enables it automatically by checking package names and signatures.

The enabled state can still be overridden on a per-package basis, but
that's not currently exposed for non-debuggable apps.

Change-Id: I86f908e227a4e2ca98940f19ba6227fbd5aaf1d1

kdrag0n committed:
gmscompat: Add overload to get signatures from ApplicationInfo

This will be used in future commits.

Change-Id: Ideb435682bcd2f24cdefc843cbfb778c8472dbb1

kdrag0n committed:
gmscompat: Add REQUEST_INSTALL_PACKAGES permission to Play Store

We have a hook to request user action for package installs from Play
Store, but it still needs the REQUEST_INSTALL_PACKAGES permission to
work.

Change-Id: I83f7ad20f90067dc68bb145fdaf5ab5a89972a92

kdrag0n committed:
gmscompat: Provide a dummy implementation of AppIntegrityManager

Newer versions of GMS crash if this service isn't available:

FATAL EXCEPTION: [com.google.android.gms.chimera.GmsIntentOperationService$ChimeraService-Executor] idle
Process: com.google.android.gms, PID: 23019
java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.content.integrity.AppIntegrityManager.getCurrentRuleSetProvider()' on a null object reference
       at com.google.android.gms.appintegrity.service.AppIntegrityModuleInitIntentOperation.e(:com.google.android.gms@212415046@21.24.15 (150400-381273646):13)
       at rpe.onHandleIntent(:com.google.android.gms@212415046@21.24.15 (150400-381273646):8)
       at com.google.android.chimera.IntentOperation.onHandleIntent(:com.google.android.gms@212415046@21.24.15 (150400-381273646):2)
       at roy.onHandleIntent(:com.google.android.gms@212415046@21.24.15 (150400-381273646):4)
       at eck.run(:com.google.android.gms@212415046@21.24.15 (150400-381273646):5)
       at ecj.run(:com.google.android.gms@212415046@21.24.15 (150400-381273646):11)
       at bvgf.run(:com.google.android.gms@212415046@21.24.15 (150400-381273646):2)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run(Thread.java:923)

Change-Id: Ic43702b78f7986d7739218fb6d0df2a26f7df270

whitequark commented:
@kdrag0n Very impressive work! I've applied this on top of Lineage 18.1 and multiple applications that require GMS work fine now.

kdrag0n committed:
gmscompat: Handle user action requests for package install sessions

This makes Play Store app installation work to an extent by handling the
user action requests required for unprivileged apps. However, Play Store
often fails to receive the "installation succeeded" event, causing it to
hang on "Installing..." until it times out.

Change-Id: Ie230cc9438706ac0668f6c21d632b375dec2ac68

kdrag0n committed:
gmscompat: Check for privileged flag instead of system

GMS apps are in /product, not /system. The privileged flag is what we
really want to check here.

Change-Id: I333c4df6ce3d02a8fa79334bd6c052d724e1a2d2

kdrag0n committed:
gmscompat: Automatically force GMS apps to be queryable

GMS apps need to be queryable from all apps by design, so set the flag
automatically instead of relying on the user to install them with `adb
install --force-queryable`.

Change-Id: I2795f3c5aac56e3cac0c024d09af4dc2a8efde7d

kdrag0n committed:
gmscompat: Skip notification posting in isolated processes

In some cases, our service hook runs in isolated processes (e.g.
sandboxed WebView services) and crashes because NotificationManager
isn't available in that context:

FATAL EXCEPTION: main
Process: com.android.webview:sandboxed_process0:org.chromium.content.app.SandboxedProcessService0:0, PID: 6722
java.lang.RuntimeException: Unable to create service org.chromium.content.app.SandboxedProcessService0: java.lang.NullPointerException: Attempt to invoke interface method 'void android.app.INotificationManager.createNotificationChannels(java.lang.String, android.content.pm.ParceledListSlice)' on a null object reference
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4201)
	at android.app.ActivityThread.access$1500(ActivityThread.java:239)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1934)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7668)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ChildZygoteInit.runZygoteServer(ChildZygoteInit.java:136)
	at com.android.internal.os.WebViewZygoteInit.main(WebViewZygoteInit.java:147)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:954)
Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'void android.app.INotificationManager.createNotificationChannels(java.lang.String, android.content.pm.ParceledListSlice)' on a null object reference
	at android.app.NotificationManager.createNotificationChannels(NotificationManager.java:834)
	at android.app.NotificationManager.createNotificationChannel(NotificationManager.java:822)
	at com.android.internal.gmscompat.GmsHooks.createFgsChannel(GmsHooks.java:83)
	at com.android.internal.gmscompat.GmsHooks.attachService(GmsHooks.java:102)
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4188)
	... 12 more

Isolated services don't need a foreground notification anyway, so just
skip the hook.

Change-Id: I0a8b161d3f2758949d02f3d28b25f9eae1b03c3c

thestinger commented:
Applied via cherry-pick due to rebase conflicts.
